Align rendering of image genercated by ImageShortCodeProvider with generic Image generation

maxime-rainville opened the pull request:
This PR aligns how ImageShortcodeProvider generates its HTML markup to be aligned with how regular images are generated.

This reduces complexity and makes it easy to customise the HTML markups.

## Parent issue
- https://github.com/silverstripe/.github/issues/230

## Depends on
- https://github.com/silverstripe/silverstripe-framework/pull/11217
  - [Review combined build](https://github.com/creative-commoners/recipe-kitchen-sink/actions/workflows/ci.yml?query=branch%3Apulls%2F5%2Falt-attribute-html)

Commit:
ENH Align rendering of image genercated by ImageShortCodeProvider with generic Image generation

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -75,10 +75,10 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
         }
 
         // Check if a resize is required
+        $manipulatedRecord = $record;
```
The old approach was basically building the `<img>` tag by hand.

I'm using standard Image manipulation to assign the correct size and attribute to a manipulated image instead.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -87,7 +87,7 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
                 $resized = $record->ResizedImage($width, $height);
                 // Make sure that the resized image actually returns an image
                 if ($resized) {
-                    $src = $resized->getURL($grant);
```
The grant isn't needed here. Session grants apply to the original file and all its variant.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -102,7 +102,7 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
             $args,
             // But enforce some values
-            ['id' => '', 'src' => $src]
+            ['id' => '', 'src' => '']
```
The `src` attirbutes need to be unset here because we want the resize picture to pick up its own URL.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -112,11 +112,14 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
 
-        $markup = self::createImageTag($attrs);
+        // We're calling renderWith() with an explicit template in case someone wants to use a custom template
+        $markup = $manipulatedRecord->renderWith(self::class . '_Image');
```
We're using a `ImageShortcodeProvider` specific template here. This would allow you to customise each template individually.

maxime-rainville reviewed (commented):
## Template tree

`templates/DBFile_image.ss` is what was used to render the mark-up for images in templates. There's a related `templates/DBFile_download.ss` that renders plain file like.

I concluded that those files were misplaced. They should be following the namespace structure.

## What I did
- Created matching `DBFile` and `DBFile_Image` templates in the correct folder.
- ImageShortcodeProvider now generates its markup by calling its own template that includes the namespaced image template.
- I moved all the mark up from the original templates and moved it to the template in the correct namespace.
  - If you've customised image rendering by overriding `templates/DBFile_image.ss`, your custom template will still be used.
  - Because the ImageShortcodeProvider doesn't call `templates/DBFile_image.ss` at all, your custom template won't start bleeding into TinyMCE images.

## Things to change in CMS 6
Once this PR is merged, I'll do a follow up PR for CMS 6:
- This [method in DBFile](https://github.com/silverstripe/silverstripe-assets/blob/ff5e8a743eea9cdddb900579a6e8c3d77a48e8ba/src/Storage/DBFile.php#L157-L171) should reference the name spaced template.
- `templates/DBFile_image.ss` and `templates/DBFile_download.ss` will be deleted.
- `ImageShortcodeProvider::createImageTag()` will be deleted since it's now deprecated.

## Other
I haven't added any test because the output of this thing is meant to be identical to the original.

I guess the only test I could run is try to override some templates and validate that I still get the expected output. Not sure how much value that would provide.

ChloeSartorelli commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -112,11 +112,14 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
 
         // Clean out any empty attributes (aside from alt) and anything not whitelisted
         $whitelist = static::config()->get('attribute_whitelist');
-        $attrs = array_filter($attrs ?? [], function ($v, $k) use ($whitelist) {
-            return in_array($k, $whitelist) && (strlen(trim($v ?? '')) || $k === 'alt');
-        }, ARRAY_FILTER_USE_BOTH);
+        foreach ($attrs as $key => $value) {
+            if (in_array($key, $whitelist) && (strlen(trim($value ?? '')) || $key === 'alt')) {
+                $manipulatedRecord = $manipulatedRecord->setAttribute($key, $value);
+            }
+        }
 
-        $markup = self::createImageTag($attrs);
+        // We're calling renderWith() with an explicit template in case someone wants to use a custom template
+        $markup = $manipulatedRecord->renderWith(self::class . '_Image');
```
We still need to call `createImageTag()` because people may have subclassed the shortcode provider, replaced the original one with theirs, and updated the `createImageTag()` method. If we stop calling it, we throw away their customisation which is a breaking change.

Instead, update the implementation of `createImageTag()` to return markup which is generated via the template. Alternatively update it to return an empty string. Then we can call it and if it returns a truthy value we know it's a custom implementation and we should use that instead of the template.

maxime-rainville replied:
On the broader point, I don't think that think semantic versionning means we will always call the same methods in the same order.

On a more narrow point, because the `createImageTag` call uses `self` rather than `static`, if you subclass `ImageShortcodeProvider` and override `createImageTag`, calling `handle_shortcode` on your subclass will still end up calling `ImageShortcodeProvider::createImageTag()`

ChloeSartorelli replied:
Ahh, I didn't catch that it uses self. I fully disagree with your reasoning for the first point but it's made moot by the use of self.

maxime-rainville commented on src/ImageManipulation.php:
```diff
@@ -1150,8 +1150,8 @@ protected function getDefaultAttributes(): array
             $attributes = [
-                'width' => $this->getWidth(),
-                'height' => $this->getHeight(),
+                'width' => $this->getWidth() ?: null,
```
When the short code was rendering an non-existent image, `getWidth` and `getHeight` would return 0. This would lead to a `width="0"` attribute in the HTML.

Defaulting to null here fixes this.

maxime-rainville commented on src/ImageManipulation.php:
```diff
@@ -1176,11 +1176,25 @@ public function getAttributes()
-            unset($attributes['loading']);
+        if (isset($attributes['loading'])) {
+            // Check if the dimensions for the image have been set
+            $dimensionsUnset = (
```
The shortcode had some logic to make sure that if no size was provided, the image would be eager loaded. This logic made more sense here in Image manipulation than in the original one.

--

The shortcode also had some logic to handle the scenario where only the height or only the width was provided. In that scenario, the missing attribute would be omitted.

Unfortunately, ImageManipulation really wants to read the size from the actual Image. This would lead to the aspect ratio of the image being changed. So I updated ImageShortcodeProvider to explicitely set the missing value to `auto` in those situation.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -75,25 +75,40 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
         $width = null;
         $height = null;
         $grant = static::getGrant($record);
-        $src = $record->getURL($grant);
```
TODO I don't need the `$src` here, but I do need to grant access to the file.

maxime-rainville replied:
Fixed.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -75,25 +75,40 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
         }
 
-        // Determine whether loading="lazy" is set
-        $args = self::updateLoadingValue($args, $width, $height);
```
ImageManipulation will take care of this from now on.

maxime-rainville replied:
TODO Delete `updateLoadingValue` down below

maxime-rainville replied:
Done.

maxime-rainville commented on tests/php/Shortcodes/ImageShortcodeProviderTest.php:
```diff
@@ -46,33 +47,31 @@ protected function tearDown(): void
-        $parser = new ShortcodeParser();
-        $parser->register('image', [ImageShortcodeProvider::class, 'handle_shortcode']);
-
-        $this->assertEquals(
-            sprintf(
-                '<img src="%s" alt="">',
-                $image->Link()
-            ),
-            $parser->parse(sprintf('[image id=%d]', $image->ID))
```
Most of these test were throwing errors because `ImageManipulation` was returning attributes in a different order.

`ImageManipulation` does a few things a bit differently which also led to slightly different result in some context.

Because none of these test actually included messages, it was really hard to figure out what each test was meant to be doing. 

So I basically ended rewriting all the test with more descriptive message and to uso a utility assert method.

maxime-rainville commented on tests/php/Shortcodes/ImageShortcodeProviderTest.php:
```diff
@@ -101,65 +100,79 @@ public function testShorcodeRegenrator()
-
-        $this->assertEquals(
-            sprintf(
-                '<img src="%s" alt="" width="50%%" height="auto">',
```
This confused me for a bit. The way to escape a `%` in `sprintf` is to double it.

maxime-rainville commented on tests/php/Shortcodes/ImageShortcodeProviderTest.php:
```diff
@@ -380,4 +482,39 @@ public function testCreateImageTag(string $expected, array $attributes)
+     * @param array $attrs Key pair values of attributes and values.
+     *                     If the value is FALSE we assume that it is not present
+     */
+    private function assertImageTag(array $attrs, string $shortcode, $message = ""): void
```
Instead of testing whether the HTML generated by a specific shortcode exactly matches the expected HTML, this method will parse the HTML and check the individual attributes.

This is better because we shouldn't care about the order of the attributes. What we care about is whether the attribute is there or not and whether it has the correct value.

maxime-rainville commented on tests/php/Shortcodes/ImageShortcodeProviderTest.php:
```diff
@@ -380,4 +482,39 @@ public function testCreateImageTag(string $expected, array $attributes)
+            $nodeAttr = $nodeAttrs->getNamedItem($key);
+            if ($value === false) {
+                if ($nodeAttr !== null) {
+                    $this->fail("$message\nImage should not contain attribute '$key'\n$tag");
```
Instead of doing straight assertion here, I decided to directly fail the test. This has a few advantages:
- When I was asserting something about the node it would give a very unpleasant about what was failing which made it difficult to figure out what was actually wrong.
- I'm building much more specific messages here:
  - I'm specifying what key is expected and what its expected value is.
  - I'm printing out the generated image tag.
  - I'm also including the overarching message coming from the original test case to better understand what's the actual scenario we are testing.

maxime-rainville commented on tests/php/Shortcodes/ImageShortcodeProviderTest.php:
```diff
@@ -380,4 +482,39 @@ public function testCreateImageTag(string $expected, array $attributes)
+                    $this->fail("$message\nImage attribute '$key' should have value '$value'\n$tag");
+                }
+            }
+            $this->assertTrue(true, 'Suppress warning about not having any assertion');
```
I need an actual assertion, otherwise PHPUnit thinks the test is broken.

maxime-rainville reviewed (commented):
Still a WIP, but here's some pointers while I have them in my mind.

The results between the old way and new way should be functionally equivalent. However, because shortcode and plain images were being rendered differently they are a couple situation where you will get slightly different results.

## No size in short code

`[img id=1]` would be converted to `<img src="..." alt="">`

Because `ImageManipulation` autodetects the size of the image the output will now be `<img src="..." alt="" width="123" height="123" loading="lazy">`,

In practice, it should not be possible to not provide sizes for images unless you bypass the UI. So this should be a very marginal use case.

In any case, I'm calling this an improvement. An image in this scenario would now be lazy loaded.

## Height without width or Width without height
`[img id=1 width="300"]` would be converted to `<img src="..." alt="" width="300">`

It will now render to `<img src="..." alt="" width="300" height="auto">`

`ImageManipulation` wants to auto-detect the size of the image if it's not provided. This could lead to a wrong aspect ratio.

So I updated the shortcode to explicitly set the missing value to `auto`. This should cause the browser to render the missing dimension to preserve the aspect ratio of the physical file.

It's worth pointing out that you can not lazy load the image in this situation because you can not predict the final size of the image.

## Rendering missing image 
ImageManipulation seem to assume that the image would always be there. It was making no effort to handle the scenario where the dimension of the image could not be determined.

I updated the logic so that if the size of an image can no be determined, ImageManipulation will now disable lazy loading on it.

In theory this could affect the scenario where you are trying to render a broken image.

maxime-rainville commented on src/Shortcodes/ImageShortcodeProvider.php:
```diff
@@ -209,29 +233,4 @@ protected static function find_error_record($errorCode)
-     * @param int|null $height
-     * @return array
-     */
-    private static function updateLoadingValue(array $args, ?int $width, ?int $height): array
```
This is private so no need to deprecate the method. We can just nuke it.

maxime-rainville commented:
The builds are red because they depend on the framework PR. I've created a branch of the sink to test this https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/9265818257

emteknetnz reviewed (changes requested):
Still have failing unit tests are merging the framework PR and rerunning

emteknetnz reviewed (changes requested):
Cannot use width/height="auto", it's invalid html

You can test that here https://www.freeformatter.com/html-validator.html

```html
<!DOCTYPE html>
<head>
<meta charset="utf-8"/>
<title>test</title>
</head>
<body>
<img src="http://example.com/abc.jpg" alt="" width="600" height="auto" />
</body>
```

`Bad value “auto” for attribute “height” on element “img”: Expected a digit but saw “a” instead.
From line 7, column 1 to line 7, column 73
Code Extract:
d>↩<body><img src="http://example.com/abc.jpg" alt="" width="600" height="auto" />↩</bo`

maxime-rainville commented:
Not sure what I was thinking there. I added a bit of logic to suppress the width/height attribute instead of outputting `auto`.

emteknetnz reviewed (changes requested):
Has the following integration failures in CI:

phpunit framework
- SilverStripe\Forms\Tests\HTMLEditor\HTMLEditorFieldTest::testReadonlyField

endtoend asset-admin
- vendor/silverstripe/asset-admin/tests/behat/features/insert-an-image.feature:108

endtoend cms
- vendor/silverstripe/cms/tests/behat/features/edit-a-page.feature:80

maxime-rainville commented:
https://github.com/silverstripe/silverstripe-framework/pull/11286 fixes the framework issue.
https://github.com/silverstripe/silverstripe-cms/pull/2971 fixes the cms issue.
My implementation was double escaping entities. I fixed it and added a unit test to cover this scenario 9d529c7a3c6793895f6da76a83b39d1a4dc814c4

emteknetnz reviewed (changes requested):
Merge conflict
